src/cryptonote_core/miner.cpp

```diff
@@ -408,5 +528,313 @@ namespace cryptonote
+    {
+      start_mining = true;
+    }
+
+    return true;
+  }
+  //-----------------------------------------------------------------------------------------------------
+  bool miner::should_i_stop_mining(const cpu_data& cpu_data_start, const cpu_data& cpu_data_end, bool& stop_mining)
+  {
+    if(!m_is_background_mining_enabled)
+    {
+      LOG_ERROR("miner::should_i_stop_mining : m_is_background_mining_enabled == false");
+      return false;
+    }
+
+    /*
+    if(m_cpu_count <= 0)
+    {
+      LOG_ERROR("miner::should_i_start_mining : m_cpu_count == 0");
+      return false;
+    }
+    */
+
+    if( !is_mining() )
+    {
+      LOG_ERROR("miner::should_i_stop_mining : is_mining() = false");
+      return false;
+    }
+
+    if( m_pausers_count > 0 )
+    {
+      LOG_ERROR("miner::should_i_stop_mining : you're not mining");
+      return false;
+    }
+
+    uint8_t idle_percentage, miner_percentage;;
+    get_cpu_idle_percentage(cpu_data_start, cpu_data_end, idle_percentage);
+    get_cpu_miner_percentage(cpu_data_start, cpu_data_end, miner_percentage);
+    if( (idle_percentage + miner_percentage) < get_idle_threshold() )
+    {
+      stop_mining = true;
+    }
+
+    return true;
+  }
+  //-----------------------------------------------------------------------------------------------------
+  void miner::get_cpu_idle_time(const cpu_data& cpu_data_start, const cpu_data& cpu_data_end, uint64_t& idle_time)
+  {
+    idle_time =
+      cpu_data_end.overall_cpu_times[CPU_DATA_OVERALL_IDLE_INDEX]
+      - cpu_data_start.overall_cpu_times[CPU_DATA_OVERALL_IDLE_INDEX];
+  }
+  //-----------------------------------------------------------------------------------------------------
+  void miner::get_cpu_active_time(const cpu_data& cpu_data_start, const cpu_data& cpu_data_end, uint64_t& active_time)
+  {
+    active_time =
+      (cpu_data_end.overall_cpu_times[CPU_DATA_OVERALL_USER_INDEX]
+      + cpu_data_end.overall_cpu_times[CPU_DATA_OVERALL_NICE_INDEX]
+      + cpu_data_end.overall_cpu_times[CPU_DATA_OVERALL_SYSTEM_INDEX])
+      -
+      (cpu_data_start.overall_cpu_times[CPU_DATA_OVERALL_USER_INDEX]
+      + cpu_data_start.overall_cpu_times[CPU_DATA_OVERALL_NICE_INDEX]
+      + cpu_data_start.overall_cpu_times[CPU_DATA_OVERALL_SYSTEM_INDEX]);
+  }
+  //-----------------------------------------------------------------------------------------------------
+  void miner::get_cpu_total_time(const cpu_data& cpu_data_start, const cpu_data& cpu_data_end, uint64_t& total_time)
+  {
+    uint64_t idle_time = 0, active_time = 0;
+    get_cpu_idle_time(cpu_data_start, cpu_data_end, idle_time);
+    get_cpu_active_time(cpu_data_start, cpu_data_end, active_time);
+    total_time = idle_time + active_time;
+  }
+  //-----------------------------------------------------------------------------------------------------
+  void miner::get_cpu_idle_percentage(const cpu_data& cpu_data_start, const cpu_data& cpu_data_end, uint8_t& idle_percentage)
```

Comment:
A number of these functions are not using any member variables. Either mark them as static, move them to the cpp entirely (remove from header), or drop the first parameter `cpu_data_start` and grab it via member function variable.